cpp/src/bitmask/bit_mask.h

```diff
@@ -0,0 +1,213 @@
+   *  @param[in] src       - the address of the source
+   *  @param[in] num_bits  - the number of bits in the bit container
+   *  @param[in] kind      - the direction of the copy
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error copy_bit_mask(bit_mask_t *dst, const bit_mask_t *src, size_t num_bits, enum cudaMemcpyKind kind) {
+    CUDA_TRY(cudaMemcpy(dst, src, num_words(num_bits) * sizeof(bit_mask_t), kind));
+    return GDF_SUCCESS;
+  }
+
+  /**
+   * Deallocate device space for the valid bitmap
+   *
+   *  @param[in]  valid   The pointer to device space that we wish to deallocate
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error destroy_bit_mask(bit_mask_t *valid) {
+    RMM_TRY(RMM_FREE(valid, 0));
+    return GDF_SUCCESS;
+  }
+
+  /**
+   * Get a single word of bits from the device
+   *
+   *  @param[out]  word - address on host where the bits will be stored
+   *  @param[out]  device_word - address on the device containing the bits to fetch
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error get_word(bit_mask_t *word, const bit_mask_t *device_word) {
+    CUDA_TRY(cudaMemcpy(word, device_word, sizeof(bit_mask_t), cudaMemcpyDeviceToHost));
+    return GDF_SUCCESS;
+  }
+	  
+  /**
+   * Put a single word of bits to the device
+   *
+   * @param[out]  word - address on host containing bits to store
+   * @param[out]  device_word - address on the device where the bits will be stored
+   */
+  inline gdf_error put_word(bit_mask_t word, bit_mask_t *device_word) {
+    CUDA_TRY(cudaMemcpy(device_word, &word, sizeof(bit_mask_t), cudaMemcpyHostToDevice));
+    return GDF_SUCCESS;
+  }
+
+  /**
+   * Allocate device space for the valid bitmap.
+   *
+   * @param[out] mask                  address of the bit mask pointer
+   * @param[in]  number_of_records     number of records
+   * @param[in]  fill_value            optional, should the memory be initialized to all 0 or 1s. All other
+   *                                   values indicate un-initialized.  Default is uninitialized
+   * @param[in]  padding               optional, specifies byte boundary the data should be padded to.
+   *                                   Defaults to 64 bytes, meaning the space allocated will be rounded
+   *                                   up to the next multiple of 64 bytes.
+   * @return GDF_SUCCESS on success, the RMM or CUDA error on error
+   */
+  gdf_error create_bit_mask(bit_mask_t **mask, gdf_size_type number_of_records, int fill_value = -1, gdf_size_type padding = 64) {
+    //
+    //  To handle padding, we will round the number_of_records up to the next padding boundary, then identify how many words
+    //  that equates to.  Then we can allocate the appropriate amount of storage.
+    //
+    gdf_size_type num_padding_blocks = (number_of_records + 8 * padding - 1) / (8 * padding);
```

Comment:
This will overflow if `number_of_records == (std::numeric_limits<gdf_size_type>::max() - 8*padding)`

Reply:
There will be less chance for overflow if you first compute the number of masks required to represent `number_of_records`, and then round the number of masks up to a multiple of the `padding_bytes`.